src/org/pentaho/reporting/platform/plugin/staging/AsyncJobFileStagingHandler.java

```
@@ -0,0 +1,189 @@
+ * See the GNU General Public License for more details.
+ *
+ *
+ * Copyright 2006 - 2016 Pentaho Corporation.  All rights reserved.
+ */
+
+package org.pentaho.reporting.platform.plugin.staging;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.pentaho.platform.api.engine.IPentahoSession;
+import org.pentaho.platform.api.util.ITempFileDeleter;
+import org.pentaho.platform.engine.core.system.PentahoSystem;
+import org.pentaho.platform.util.UUIDUtil;
+import org.pentaho.reporting.engine.classic.core.util.StagingMode;
+import org.pentaho.reporting.platform.plugin.TrackingOutputStream;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * Async stage handler.
+ * Write to TEMP file but:
+ * - do not keep link to passed output stream not write to output stream passed in constructor.
+ * - live between requests.
+ * - require to re-set output stream for ready-to-fetch-request to fetch ready to use data.
+ *
+ *
+ *
+ * Created by dima.prokopenko@gmail.com on 2/10/2016.
+ */
+public class AsyncJobFileStagingHandler extends AbstractStagingHandler {
+
+  private static String POSTFIX = ".tmp";
+
+  public static final String STAGING_DIR_ATTR = "asyncstaging";
+
+  private static final Log logger = LogFactory.getLog( AsyncJobFileStagingHandler.class );
+
+  private TrackingOutputStream fileTrackingStream;
+  // package private for testing purpose
+  File tmpFile;
+
+  private static String stringParentDir = ".";
+
+  public AsyncJobFileStagingHandler( OutputStream outputStream, IPentahoSession userSession, String stringParentDir ) throws IOException {
```

If you dont use the output stream, then dont pass it in. The only time this class is instantiated is in a manual constructor call - so delete that parameter. 

Initialize() will fail if stringParentDir is null. This variable is only ever set in the constrcutor. So do that check here, not in initialize(), and thus fail early!